src/libs/decoders/flac.c

```diff
@@ -0,0 +1,184 @@
+    const int whence = (origin == drflac_seek_origin_start) ? RW_SEEK_SET : RW_SEEK_CUR;
+    Sound_Sample *sample = (Sound_Sample *) pUserData;
+    Sound_SampleInternal *internal = (Sound_SampleInternal *) sample->opaque;
+    return (SDL_RWseek(internal->rw, offset, whence) != -1) ? DRFLAC_TRUE : DRFLAC_FALSE;
+} /* flac_seek */
+
+
+static int FLAC_init(void)
+{
+    return 1;  /* always succeeds. */
+} /* FLAC_init */
+
+
+static void FLAC_quit(void)
+{
+    /* it's a no-op. */
+} /* FLAC_quit */
+
+static int FLAC_open(Sound_Sample *sample, const char *ext)
+{
+    Sound_SampleInternal *internal = (Sound_SampleInternal *) sample->opaque;
+    drflac *dr = drflac_open(flac_read, flac_seek, sample, NULL);
+
+    if (!dr)
+    {
+        BAIL_IF_MACRO(sample->flags & SOUND_SAMPLEFLAG_ERROR, ERR_IO_ERROR, 0);
+        BAIL_MACRO("FLAC: Not a FLAC stream.", 0);
+    } /* if */
+
+    SNDDBG(("FLAC: Accepting data stream.\n"));
+    sample->flags = SOUND_SAMPLEFLAG_CANSEEK;
+
+    sample->actual.channels = dr->channels;
+    sample->actual.rate = dr->sampleRate;
+    sample->actual.format = AUDIO_S16SYS; /* returns native byte-order based on architecture */
+
+    const Uint64 frames = (Uint64) dr->totalPCMFrameCount;
+    if (frames == 0)
+        internal->total_time = -1;
+    else
+    {
+        const Uint32 rate = (Uint32) dr->sampleRate;
+        internal->total_time = ( (Sint32)frames / rate) * 1000;
+        internal->total_time += ((frames % rate) * 1000) / rate;
+    } /* else */
+
+    internal->decoder_private = dr;
+
+    return 1;
+} /* FLAC_open */
+
+
+
+static void FLAC_close(Sound_Sample *sample)
+{
+    Sound_SampleInternal *internal = (Sound_SampleInternal *) sample->opaque;
+    drflac *dr = (drflac *) internal->decoder_private;
+    drflac_close(dr);
+} /* FLAC_close */
+
+
+static Uint32 FLAC_read(Sound_Sample *sample)
+{
+    Sound_SampleInternal *internal = (Sound_SampleInternal *) sample->opaque;
+    drflac *dr = (drflac *) internal->decoder_private;
+    const drflac_uint64 rc = drflac_read_pcm_frames_s16(dr,
+                                                        internal->buffer_size / (dr->channels * sizeof(drflac_int16)),
+                                                        (drflac_int16 *) internal->buffer);
+    return (Uint32) rc * dr->channels * sizeof (drflac_int16);
+} /* FLAC_read */
+
+
+static int FLAC_rewind(Sound_Sample *sample)
+{
+    Sound_SampleInternal *internal = (Sound_SampleInternal *) sample->opaque;
+    drflac *dr = (drflac *) internal->decoder_private;
+    return (drflac_seek_to_pcm_frame(dr, 0) == DRFLAC_TRUE);
+} /* FLAC_rewind */
+
+static int FLAC_seek(Sound_Sample *sample, Uint32 ms)
+{
+    Sound_SampleInternal *internal = (Sound_SampleInternal *) sample->opaque;
+    drflac *dr = (drflac *) internal->decoder_private;
+    const float frames_per_ms = ((float) sample->actual.rate) / 1000.0f;
+    const drflac_uint64 frame_offset = (drflac_uint64) (frames_per_ms * ((float) ms));
```

Comment:
Yes; the truncation is desired, but to be more explicit I replaced these with llroundf's (flac and wav) and lroundf (vorbis).  

The round functions return a signed int, but we store them in an unsigned value because frames are always positive - so we'll see if that throws more warnings. Worst case, I can explicitly cast the signed to unsigned.

There are a lot more floating-point to integer conversions in the code (all of them truncate down expect the +0.5 one you flagged); does it make sense or is more appropriate to wrap them all with respective round(...) calls?  If it's considered better, I'm am happy to do it.

Reply:
Fixed in https://github.com/dreamer/dosbox-staging/commit/02a974f80405e5cbd687a894cf9930e2a0c251ad

Reply:
Rounding in wav.c fixed in https://github.com/dreamer/dosbox-staging/commit/02a974f80405e5cbd687a894cf9930e2a0c251ad